### ISSUE TITLE
Add op fails with broadcasting when inputs are int32

### ISSUE THREAD

**Issue by sshonTT** [tenstorrent/tt-metal#21480]
### Describe the bug

It seems that Add op in TTNN does not correctly handle broadcasting when inputs are of int32 type.

This issue is particularly noticeable when an Expand op with int32 inputs is lowered to an Add with mismatched shapes. Although TTNN claims to support int32 for elementwise ops like Add and Subtract, broadcasting fails silently or produces incorrect results in this case.

### Steps to reproduce the issue

in tt-torch workspace,

import torch
from torch import nn
import pytest
import math
import threading

import tt_torch
from tt_torch.tools.verify import verify_module
from tt_torch.tools.utils import CompilerConfig
from tt_torch.tools.device_manager import DeviceManager
from torch_mlir import fx
from tt_torch.dynamo.backend import backend

from torch_mlir.compiler_utils import (
    OutputType,
    run_pipeline_with_repro_report,
    lower_mlir_module,
)

def test_inc():
    class Basic(nn.Module):
        def __init__(self):
            super().__init__()

        def forward(self, b):
            return b.expand(1, 1, 256, 256)

    a = torch.cat([
        torch.zeros(128, dtype=torch.int32),
        torch.ones(128, dtype=torch.int32)
    ])
    a = a.reshape(1, 1, 1, 256)  # final shape: (1, 1, 1, 256)
    mod = fx.export_and_import(Basic(), a, output_type=OutputType.STABLEHLO)
    mod.dump()
    tt_mod = torch.compile(Basic(), backend=backend)
    out = tt_mod(a)
    print(f"out: {out}")

if __name__ == "__main__":
    test_inc()

### Expected behavior

All values are correctly expanded.

**Comment by KalaivaniMCW** [comment]
should be working with binary_ng

**Comment by sshonTT** [comment]
@KalaivaniMCW I don’t have an issue with expand in TTNN itself, but I do run into a problem when trying to run add with broadcasting using int32 dtype.
I’ve included a repro below:

```
def test_add_i32(device):
    torch.manual_seed(2024)
    a = torch.cat([
        torch.zeros(128, dtype=torch.int32),
        torch.ones(128, dtype=torch.int32)
    ])
    a = torch.reshape(a, (1,1,1,256))
    b = torch.zeros((1,1,256,256))

    torch_add = a + b

    input_tensor_a = ttnn.from_torch(a, layout=ttnn.TILE_LAYOUT, device=device)
    input_tensor_b = ttnn.from_torch(b, layout=ttnn.TILE_LAYOUT, device=device)
    output_tensor = ttnn.add(input_tensor_a, input_tensor_b)

    output_tensor = ttnn.to_torch(output_tensor)
    assert torch.allclose(torch_add, output_tensor, atol=1e-1, rtol=1e-2)
```
Now that torch.expand is decomposed into a broadcasting add operation in TT-MLIR, this should be a more precise repro case.

**Comment by KalaivaniMCW** [comment]
Hi @sshonTT 

1. the broadcast support for add is not enabled by default but it can be enabled by setting `use_legacy=False` 
2. the broadcast functionality will be available by default soon. 
I made minor modifications here in the test - set the input tensor dtype & switched the assert to equal since we are comparing integers.
Test passes.

```
def test_add_i32(device):
    torch.manual_seed(2024)
    a = torch.cat([
        torch.zeros(128, dtype=torch.int32),
        torch.ones(128, dtype=torch.int32)
    ])
    a = torch.reshape(a, (1,1,1,256))
    b = torch.zeros((1,1,256,256))

    torch_add = a + b

    input_tensor_a = ttnn.from_torch(a, dtype=ttnn.int32, layout=ttnn.TILE_LAYOUT, device=device)
    input_tensor_b = ttnn.from_torch(b, dtype=ttnn.int32, layout=ttnn.TILE_LAYOUT, device=device)
    output_tensor = ttnn.add(input_tensor_a, input_tensor_b, use_legacy=False)

    output_tensor = ttnn.to_torch(output_tensor)
    assert torch.equal(torch_add, output_tensor)
```

<img width="1050" alt="Image" src="https://github.com/user-attachments/assets/6174b7d8-b248-4245-9f66-c6c228631326" />

**Comment by sshonTT** [comment]
@KalaivaniMCW Thanks! I'm not sure how tt-rt is calling add with the use_legacy field at the moment, but I hope the broadcasting issue gets fixed soon.

**Comment by cmaryanTT** [comment]
Binary_ng was merged last week, this should be working now without use_legacy=false - just remove that flag entirely.

**Comment by KalaivaniMCW** [comment]
Still depends on `use_legacy=False` flag, as broadcast happens on H-dim which causes to fallback to legacy. 
Broadcasting for binary ops can be enabled by passing the flag `ttnn_op(input_tensor_a, input_tensor_b, use_legacy=False)`

**Comment by cmaryanTT** [comment]
@dchenTT - the issue here is that int32 isn't supported for row broadcast in the old OP (perhaps also other dtypes and other broadcast dims). Need to adjust the fallback to use binary_ng in these cases, even though perf isn't great.

**Comment by dchenTT** [comment]
By now we have a lot of binary ops with broadcasting behavior that only works on binary_ng, not legacy. While we internally have a heuristic to fallback to legacy automatically, but the cases cannot be exhausted easily. In this case, the recommend behavior is to set flag use_legacy=False. Only when that does not work, then consider as a real bug. Eventually the heuristic will cover all cases or no heuristic that everything is binary_ng.

**Comment by dchenTT** [comment]
With this PR https://github.com/tenstorrent/tt-metal/pull/24255 integer binary op has been directly to binary_ng, unless explicated by use_legacy=True. So this issue is considered fixed.

**Comment by dchenTT** [comment]
Please verify and close.
